File: TerminalVoiceOverlay-Termux/src/gemini-client.ts
import https from "node:https";

const RETRYABLE_STATUS = new Set([429, 500, 503]);
const MAX_RETRIES = 5;
const DELAYS = [2000, 4000, 8000, 16000, 32000];

const SYSTEM_PROMPT = `ROLLE:
Du bist ein technischer Redakteur, spezialisiert auf die Aufbereitung von Spracheingaben für KI-Coding-Tools. Du verstehst Programmierkonzepte und bewahrst technische Präzision, während du gesprochene Sprache in klare schriftliche Anweisungen umwandelst.

AUFGABE:
Du erhältst einen diktierten Text (Speech-to-Text). Der Sprecher spricht Deutsch, verwendet aber regelmäßig englische Fachbegriffe aus der Programmierung (Funktionsnamen, Frameworks, CLI-Befehle etc.). Die Spracherkennung kann diese englischen Begriffe falsch transkribieren – erkenne und korrigiere solche Fehler anhand des technischen Kontexts. Der Sprecher gibt Programmier-Anweisungen an ein KI-Coding-Tool (z.B. Claude Code). Bereite den Text so auf, dass er als klare, präzise Eingabe funktioniert.

VORGEHEN:
1) Entferne Diktat-Artefakte: Fülllaute ("äh", "ähm"), Stotterer, Wortwiederholungen, sinnlose Fragmente.
2) Erkenne und korrigiere falsch transkribierte englische Fachbegriffe (z.B. "use Tate" → "useState", "Fötch" → "fetch").
3) Erkenne die Absicht und formuliere den Text als klare Anweisung um. Sätze dürfen umstrukturiert, Wortwahl verbessert und Satzgrenzen neu gesetzt werden. Zusammengehörige Anweisungen als einen Befehl belassen.
4) Korrigiere Grammatik, Zeichensetzung und Groß-/Kleinschreibung.
5) Bewahre technische Begriffe EXAKT: Dateinamen, Funktionsnamen, Variablen, Programmiersprachen, Frameworks, CLI-Befehle, API-Namen – NICHT übersetzen oder verändern.

GRENZEN:
- Keine neuen Informationen oder Vermutungen hinzufügen.
- Intention des Originals vollständig erhalten.
- Englische Fachbegriffe und Code-Referenzen NIEMALS eindeutschen.
- Sprache: Deutsch (außer technische Begriffe).

AUSGABEFORMAT:
- Ausschließlich den überarbeiteten Text. Keine Kommentare, keine Erklärungen, kein Präfix.
- Ausführliche, vollständige Sätze, sodass jede Intention des Sprechers klar und unmissverständlich beim Leser ankommt.
- Natürlicher, verständlicher Ton – so wie man einem Kollegen etwas erklärt. Kein Behördendeutsch, keine Geschäftsbrief-Floskeln, keine gestelzte Sprache.`;

export class GeminiClient {
  private apiKey: string;
  private model: string;

  constructor(apiKey: string, model = "gemini-3.1-flash-lite-preview") {
    this.apiKey = apiKey;
    this.model = model;
  }

  async correctText(text: string): Promise<string> {
    const prompt = `${SYSTEM_PROMPT}\n\nTEXT_START\n${text}\nTEXT_END`;
    return this.sendRequest(prompt, 0);
  }

  private async sendRequest(prompt: string, attempt: number): Promise<string> {
    const payload = JSON.stringify({
      contents: [{ role: "user", parts: [{ text: prompt }] }],
      generationConfig: {
        maxOutputTokens: 8192,
        thinkingConfig: { thinkingLevel: "MEDIUM" },
      },
    });

    const path = `/v1beta/models/${this.model}:generateContent?key=${this.apiKey}`;

    return new Promise((resolve, reject) => {
      const req = https.request(
        {
          hostname: "generativelanguage.googleapis.com",
          path,
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "Content-Length": Buffer.byteLength(payload),
          },
          timeout: 120000,
        },
        (res) => {
          const chunks: Buffer[] = [];
          res.on("data", (chunk: Buffer) => chunks.push(chunk));
          res.on("end", () => {
            const responseBody = Buffer.concat(chunks).toString("utf-8");
            const statusCode = res.statusCode || 0;

            if (statusCode < 200 || statusCode >= 300) {
              if (RETRYABLE_STATUS.has(statusCode) && attempt < MAX_RETRIES) {
                const delay = DELAYS[attempt]!;
                setTimeout(() => {
                  this.sendRequest(prompt, attempt + 1).then(resolve, reject);
                }, delay);
                return;
              }
              reject(new Error(`Gemini API Fehler ${statusCode}: ${responseBody}`));
              return;
            }

            try {
              const text = this.extractText(responseBody);
              resolve(text);
            } catch (e) {
              reject(e);
            }
          });
        }
      );

      req.on("error", reject);
      req.on("timeout", () => {
        req.destroy();
        reject(new Error("Gemini API Timeout (120s)"));
      });
      req.write(payload);
      req.end();
    });
  }

  private extractText(body: string): string {
    const json = JSON.parse(body);
    const parts = json?.candidates?.[0]?.content?.parts;
    if (!Array.isArray(parts)) throw new Error("Unerwartete Gemini-Antwort");

    for (const part of parts) {
      if (part.thought) continue;
      if (part.text) return part.text.trim();
    }
    throw new Error("Kein Text in Gemini-Antwort");
  }
}

File: ClaudeCodexVoiceOverlay-macOS/ClaudeCodexVoiceOverlay/GeminiClient.swift
import Foundation

final class GeminiClient {
    private let apiKey: String
    private let model = "gemini-3.1-flash-lite-preview"
    private let retryableStatusCodes: Set<Int> = [429, 500, 503]
    private let maxRetries = 5
    private let delays: [TimeInterval] = [2, 4, 8, 16, 32]

    init(apiKey: String) {
        self.apiKey = apiKey
    }

    /// Pfad zur externen Gemini-Korrektur-Prompt-Datei. Wird bei JEDEM
    /// correctText-Aufruf neu gelesen — Aenderungen wirken sofort ohne
    /// App-Neustart. Fehlt die Datei: Fallback auf eingebauten Default-Prompt.
    private static var geminiPromptFilePath: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("SK/VoiceOverlays/gemini-correction-prompt.txt")
    }

    private static func loadCorrectionPrompt() -> String? {
        let url = geminiPromptFilePath
        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    func correctText(_ text: String, completion: @escaping (Result<String, Error>) -> Void) {
        // Externe Korrektur-Prompt-Datei hat Vorrang — Frank kann sie
        // jederzeit pflegen ohne Rebuild. Fallback: eingebauter Default.
        let template = Self.loadCorrectionPrompt() ?? GeminiClient.defaultCorrectionTemplate
        let prompt = template + "\n" + text + "\nTEXT_END"

        DispatchQueue.global(qos: .userInitiated).async { [self] in
            self.sendRequest(prompt: prompt, attempt: 0, completion: completion)
        }
    }

    private static let defaultCorrectionTemplate = """
    ROLLE:
    Du bist ein technischer Redakteur, spezialisiert auf die Aufbereitung von Spracheingaben für KI-Coding-Tools. Du verstehst Programmierkonzepte und bewahrst technische Präzision, während du gesprochene Sprache in klare schriftliche Anweisungen umwandelst.

    AUFGABE:
    Du erhältst einen diktierten Text (Speech-to-Text). Der Sprecher spricht Deutsch, verwendet aber regelmäßig englische Fachbegriffe aus der Programmierung. Die Spracherkennung kann diese englischen Begriffe falsch transkribieren – erkenne und korrigiere solche Fehler anhand des technischen Kontexts.

    VORGEHEN:
    1) Entferne Diktat-Artefakte: Fülllaute, Stotterer, Wortwiederholungen.
    2) Erkenne und korrigiere falsch transkribierte englische Fachbegriffe.
    3) Erkenne die Absicht und formuliere den Text als klare Anweisung um.
    4) Korrigiere Grammatik, Zeichensetzung und Groß-/Kleinschreibung.
    5) Bewahre technische Begriffe EXAKT.

    AUSGABEFORMAT:
    Ausschließlich den überarbeiteten Text. Keine Kommentare.

    TEXT_START
    """

    private func sendRequest(prompt: String, attempt: Int, completion: @escaping (Result<String, Error>) -> Void) {
        var urlComponents = URLComponents(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent")!
        urlComponents.queryItems = [URLQueryItem(name: "key", value: apiKey)]

        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 120

        let payload: [String: Any] = [
            "contents": [
                ["role": "user", "parts": [["text": prompt]]]
            ],
            "generationConfig": [
                "maxOutputTokens": 8192,
                "thinkingConfig": ["thinkingLevel": "MEDIUM"]
            ] as [String: Any]
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        let task = URLSession.shared.dataTask(with: request) { [self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0

            if !(200...299).contains(statusCode) {
                if self.retryableStatusCodes.contains(statusCode) && attempt < self.maxRetries {
                    let delay = self.delays[attempt]
                    NSLog("Gemini %d - retry %d/%d, waiting %.0fs...", statusCode, attempt + 1, self.maxRetries, delay)
                    DispatchQueue.global(qos: .userInitiated).asyncAfter(deadline: .now() + delay) {
                        self.sendRequest(prompt: prompt, attempt: attempt + 1, completion: completion)
                    }
                    return
                }
                let responseText = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                completion(.failure(APIError.httpError(statusCode, responseText)))
                return
            }

            guard let data = data else {
                completion(.failure(APIError.noData))
                return
            }

            do {
                let text = try self.extractText(from: data)
                completion(.success(text))
            } catch {
                completion(.failure(error))
            }
        }
        task.resume()
    }

    private func extractText(from data: Data) throws -> String {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let candidates = json["candidates"] as? [[String: Any]],
              let content = candidates.first?["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]] else {
            throw APIError.unexpectedResponse
        }

        // Skip thinking parts (thought: true), use only text parts
        for part in parts {
            if let isThought = part["thought"] as? Bool, isThought { continue }
            if let text = part["text"] as? String, !text.isEmpty {
                return text.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        throw APIError.noTextInResponse
    }

    enum APIError: Error, LocalizedError {
        case httpError(Int, String)
        case noData
        case unexpectedResponse
        case noTextInResponse

        var errorDescription: String? {
            switch self {
            case .httpError(let code, let msg): return "Gemini API Fehler \(code): \(msg)"
            case .noData: return "Keine Daten von Gemini erhalten"
            case .unexpectedResponse: return "Unerwartete Gemini-Antwort"
            case .noTextInResponse: return "Kein Text in Gemini-Antwort"
            }
        }
    }
}
